api-docs: Move api docs to top level directory.

laurynmm opened the pull request:
Moves the API documentation markdown files from `templates/zerver/api` to top level `api_docs/` directory. Follow-up to #24038, which made a similar change to the help documentation.

---

**Notes**:
- 1st commit is a prep commit to update where we check for `api-doc-template.md` being called directly and return 404 / missing for that case. And adds a test for that in `zerver/tests/test_urls`.
- 3 following commits should be squashed before merging, but separate now to help with review/update process.
  - 2nd commit: Moves the files to the new directory.
  - 3rd commit: Updates documentation for new directory.
  - 4th commit: Updates the backend code for rendering the API documentation.

After these updates, the only relative path cases for help, API or policies documentation should be when an organization has not set up their policies documentation (`templates/zerver/policies_absent`) or the policies documentation in the development environment (`templates/corporate/policies`). Because these are in the `templates` directory, they are still processed through the jinja template engines in `render_markdown_path`.

Ran a diff for the API documentation with these changes against the current main branch in the dev environment, and saw no changes.

---
 
**Follow-ups/Related PRs**:
- In `zerver/lib/templates.py`, we pass `templates/zerver/api` as the value for `base_path` to the markdown extensions for API endpoint parameters and return values. I'm planning on cleaning those cases up in a separate pull request. They don't impact the current API documentation because everything is being rendered from `zerver/openapi/zulip.yaml`. The `base_path` was previously used when endpoint parameters were documented in JSON files.

---

<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [x] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [ ] Explains differences from previous plans (e.g., issue description).
- [x] Highlights technical choices and bugs encountered.
- [x] Calls out remaining decisions and concerns.
- [x] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [ ] Each commit is a coherent idea.
- [ ] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [ ] Visual appearance of the changes.
- [ ] Responsiveness and internationalization.
- [ ] Strings and tooltips.
- [ ] End-to-end functionality of buttons, interactions and flows.
- [ ] Corner cases, error conditions, and easily imagined bugs.
</details>

timabbott commented on tools/linter_lib/custom_check.py:
```diff
@@ -842,17 +844,20 @@
         {
             "pattern": "su zulip -c [^']",
+            "exclude": {"api_docs/"},
             "include_only": {"docs/"},
```
This has an `include_only`, so I don't think you need the exclude. Same with the next two.

laurynmm replied:
Looking at the documentation comment for `include_only` in [zulint custom_rules.py](https://github.com/zulip/zulint/blob/80aba3fd88faf6598665ac670a1d30f3c028df0d/zulint/custom_rules.py#L165):
```python
        # 'include_only': 'set([<path>, ...])' - includes only those files where <path> is a
        #                                        substring of the filepath.
```

So, `"docs/"` is a substring of `"api_docs/"` hence the custom rules here needed to be updated.

This was caught when I opened the PR and CI was failing on the backend lint step. I put up the current change quickly as it was the end of my day, so it wasn't well thought through as a solution. Will look more at it today.

timabbott commented on zerver/views/documentation.py:
```diff
@@ -91,7 +91,7 @@ def get_path(self, article: str) -> DocumentationArticle:
 
-        # Absolute path cases
+        # Absolute path cases for help and policies documentation
         if (self.policies_view or self.help_view) and self.path_template.startswith("/"):
```
Does the `path_template.startswith` check here do anything? I ask mainly because it makes it feel pretty hard to think about in what situations the final else statement below will apply.

laurynmm replied:
It did do something here because there are cases where the policies documentation may be in `templates/`, which will be a relative path. I updated this `get_path` function to consider that specific case first with a comment so that it's a bit clearer why we're checking for relative paths and loading templates.

timabbott commented on zerver/views/documentation.py:
```diff
@@ -105,22 +105,36 @@ def get_path(self, article: str) -> DocumentationArticle:
-                        endpoint_path=None,
-                        endpoint_method=None,
-                    )
+        if os.path.exists(path):
```
Alternatively, could this if/else block have an `and self.api_view` or something both here and in the next `elif`?

laurynmm replied:
I added the boolean `self.api_doc_view` to use for these checks, but also reworked the flow of this function overall in the updated version.

timabbott commented on zerver/views/documentation.py:
```diff
@@ -144,17 +158,11 @@ def get_context_data(self, **kwargs: Any) -> Dict[str, Any]:
-        if documentation_article.article_path.startswith("/") and os.path.exists(
-            documentation_article.article_path
-        ):
+        if os.path.exists(documentation_article.article_path):
```
Why is the `startswith` removed here? The comment just below says this is the absolute path case.

One thing we should be careful about in this code path is security; we want to make sure you can't visit a URL like zulip.example.com/help/../../../../etc/zulip/zulip-secrets.conf; which I'm pretty sure is enforced mainly through the `urls.py` configuration.

laurynmm replied:
I added the `startswith` check back in.

The `urls.py` configuration for api and policies views are both `slug:article`, so that is enforced there.

Your example above for help views happens in `get_path`, when we set article to be "missing" for any case other than "include/sidebar_index" ...
```python
        elif article == "include/sidebar_index":
            pass
        elif "/" in article:
            article = "missing"
            http_status = 404
```

timabbott commented on zerver/views/documentation.py:
```diff
@@ -187,17 +195,15 @@ def get_context_data(self, **kwargs: Any) -> Dict[str, Any]:
             with open(article_absolute_path) as article_file:
                 first_line = article_file.readlines()[0]
-            if context["article"] == "/zerver/api/api-doc-template.md":
+            if context["article"].endswith("api-doc-template.md"):
```
I think an `if api_docs_view and ...` boolean here would be helpful for readability... and then code above can do things like `else: raise AssertionError("Invalid documentation view type")` rather than having the API docs just be an `else` clause above.

laurynmm replied:
This made sense to me, so each type of documentation view now has a boolean set to 'True'.

laurynmm commented on zerver/views/documentation.py:
```diff
@@ -91,50 +92,53 @@ def get_path(self, article: str) -> DocumentationArticle:
+        if not self.path_template.startswith("/"):
+            # Relative paths only used for policies documentation
+            # when it is not configured or in the dev environment
+            assert self.policies_view
```
I added this assertion here, but wasn't sure if it was necessary.

laurynmm commented:
**Updates**:
- Added `api_doc_view` boolean to `MarkdownDirectoryView`. Now all three doc types have a boolean value that's used in `get_path` and `get_context_data`.
- Reworked how `get_path` in `MarkdownDirectoryView` goes through the different paths (relative / absolute). Now the relative paths for policies documentation are checked / returned first.
- Only excluded `api_docs/` directory from one lint rule for relative links, see [comment above](https://github.com/zulip/zulip/pull/24241#discussion_r1093474175).

timabbott commented:
Squashed the last 3 commits and merged, thanks for doing this migration @laurynmm!

The only thing that's now a bit odd about this setup is something that's kinda been true for a while, namely that this directory only contains the non-REST API documentation pages, and that's not super discoverable. 

We should discuss in #documentation, but one possible answer is that `zerver/openapi/` should be moved to `api_docs/openapi`, or some other architecture that puts these in the same tree.
